File: src/python/problems/predict_party_victory_test.py
#https://leetcode.com/problems/dota2-senate/

from core.problem_base import *

class PredictPartyVictory(ProblemBase):
    def Solution(self, senate: str) -> str:
        voting = []
        radiantTotal = 0
        radiantSkip = 0
        direTotal = 0
        direSkip = 0

        for senator in senate:
            voting.append(senator)
            if senator == 'R':
                radiantTotal += 1
            else:
                direTotal += 1
        
        while radiantTotal > 0 and direTotal > 0:
            senator = voting.pop(0)
            if senator == 'R':
                if radiantSkip <= 0:
                    direSkip += 1
                    direTotal -= 1
                    voting.append(senator)
                else:
                    radiantSkip -= 1
            elif direSkip <= 0:
                radiantSkip += 1
                radiantTotal -= 1
                voting.append(senator)
            else:
                direSkip -= 1

        return "Radiant" if direTotal == 0 else "Dire"
    
    def Solution1(self, senate: str) -> str:
        radiant = []
        dire = []
        
        for order in range(len(senate)):
            if senate[order] == 'R':
                radiant.append(order)
            else:
                dire.append(order)
        
        while radiant and dire:
            radiantVoter = radiant.pop(0)
            direVoter = dire.pop(0)
            if radiantVoter < direVoter:
                radiant.append(radiantVoter + len(senate))
            else:
                dire.append(direVoter + len(senate))

        return "Radiant" if radiant else "Dire"

if __name__ == '__main__':
    TestGen(PredictPartyVictory) \
        .Add(lambda tc: tc.Param("RD").Result("Radiant")) \
        .Add(lambda tc: tc.Param("RDD").Result("Dire")) \
        .Add(lambda tc: tc.Param("RRDDD").Result("Radiant")) \
        .Add(lambda tc: tc.Param("DRRDRDRDRDDRDRDR").Result("Radiant")) \
        .Run()

File: src/challenges/go/problems/predict_party_victory_test.go
// https://leetcode.com/problems/dota2-senate

package problems

import (
	"testing"
	"github.com/dobermanch/leetcode/core"
)

type PredictPartyVictory struct{}

func TestPredictPartyVictory(t *testing.T) {
	gen := core.TestSuite[PredictPartyVictory]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("RD").Result("Radiant")
	}).Add(func(tc *core.TestCase) {
		tc.Param("RDD").Result("Dire")
	}).Add(func(tc *core.TestCase) {
		tc.Param("RRDDD").Result("Radiant")
	}).Add(func(tc *core.TestCase) {
		tc.Param("DRRDRDRDRDDRDRDR").Result("Radiant")
	}).Run(t)
}

func (PredictPartyVictory) Solution(senate string) string {
	voting := []rune{};
	radiantTotal := 0
    radiantSkip := 0
	direTotal := 0
    direSkip := 0

	for _,senator := range senate {
		if senator == 'R' {
			radiantTotal++
		} else {
			direTotal++
		}

		voting = append(voting, senator)
	}

	for radiantTotal > 0 && direTotal > 0 {
		if voting[0] == 'R' {
			if radiantSkip <= 0 {
				direSkip++
				direTotal--
				voting = append(voting, voting[0])
			} else {
				radiantSkip--
			}
		} else if direSkip <= 0 {
			radiantSkip++
			radiantTotal--
			voting = append(voting, voting[0])
		} else {
			direSkip--
		}   

		voting = voting[1:]
	}

	if direTotal == 0 {
		return "Radiant"
	} else { 
		return "Dire"
	}
}

func (PredictPartyVictory) Solution1(senate string) string {
	radiant := []int{}
    dire := []int{}

    for order, senator := range senate {
        if senator == 'R' {
            radiant = append(radiant, order)
        } else {
            dire = append(dire, order)
        }
    }

    for len(radiant) > 0 && len(dire) > 0 {
        if radiant[0] < dire[0] {
            radiant = append(radiant, radiant[0] + len(senate))
        } else {
            dire = append(dire, dire[0] + len(senate))
        }

        radiant = radiant[1:]
        dire = dire[1:]
    }

    if len(radiant) > 0 {
        return "Radiant"
    } else {
        return "Dire"
    }
}
